### ISSUE TITLE
[C++][R][Packaging] r-binary-packages macOS High Sierra failed by BMI2 related error

### ISSUE THREAD

**Issue by kou** [apache/arrow#37017]
### Describe the bug, including details regarding any error messages, version, and platform.

The last success nightly build (2023-07-19): https://github.com/ursacomputing/crossbow/actions/runs/5596846018/job/15162699322
The first failed nightly build (2023-07-20): https://github.com/ursacomputing/crossbow/actions/runs/5608731289/job/15196725675#step:13:7156

```text
/Users/voltrondata/tmp/hbtmp/apache-arrow-20230720-60391-l8lhul/cpp/src/arrow/compute/key_map_avx2.cc:132:30: error: always_inline function '_pext_u32' requires target feature 'bmi2', but would be inlined into function 'early_filter_imp_avx2_x8' that is compiled without support for 'bmi2'
    out_match_bitvector[i] = _pext_u32(_mm256_movemask_epi8(vmatch_found),
                             ^
```

I think that #36662 is related.

### Component(s)

C++, Packaging, R

**Comment by thisisnic** [comment]
CC @pitrou

**Comment by pitrou** [comment]
Hmm, do we use BMI2 instructions in AVX2-optimized routines? That seems wrong.
@westonpace @felipecrv

**Comment by felipecrv** [comment]
> Hmm, do we use BMI2 instructions in AVX2-optimized routines? That seems wrong. @westonpace @felipecrv

The use of `_pext_u32` in `key_map_vx2.cc` dates back to 2021. Given that both AVX2 and BMI2 were introduced at the same time (Haswell), would it be a problem to also enable BMI2 when we enable AVX2 in the build?

**Comment by thisisnic** [comment]
I think this is the cause of #37596; any idea what we should do here @pitrou ?

**Comment by pitrou** [comment]
@thisisnic I can try to fix the issue on the C++ side.

**Comment by pitrou** [comment]
Hmm, actually, we already seem to consider that compilers supporting AVX2 also support BMI2. That leaves the issue of adding the proper compile flags, and also of fixing runtime dispatch.

**Comment by pitrou** [comment]
By the way, is there a reason the macOS package builds in `r-binary-packages` depends on the `windows-cpp` build here? https://github.com/apache/arrow/blob/48cc2abe29fb261317d676090931b245e82f136e/dev/tasks/r/github.packages.yml#L138

@assignUser

**Comment by assignUser** [comment]
@pitrou that step is a matrix job that builds the r package for mac and windows, as there are no prebuild mac binaries (at the moment at least^^) we need to wait for the windows step. There is no real lost time here though as we need to wait for both steps to finish before we can upload the packages so :shrug: it's not worth the duplicated code to  split it into two jobs. (well if I have to test mac only changes repeatedly, as you are here, I sometimes remove the windows part of the  job to speed up test times)
